pkg/block/indexheader/binary_reader.go

```diff
@@ -0,0 +1,817 @@
+	p := fw.pos % uint64(size)
+	if p == 0 {
+		return nil
+	}
+	p = uint64(size) - p
+
+	if err := fw.Write(make([]byte, p)); err != nil {
+		return errors.Wrap(err, "add padding")
+	}
+	return nil
+}
+
+func (fw *FileWriter) Close() error {
+	if err := fw.Flush(); err != nil {
+		return err
+	}
+	if err := fw.f.Sync(); err != nil {
+		return err
+	}
+	return fw.f.Close()
+}
+
+func (fw *FileWriter) Remove() error {
+	return os.Remove(fw.name)
+}
+
+func (w *binaryWriter) AddIndexMeta(indexVersion int, indexPostingOffsetTable uint64) error {
+	w.buf.Reset()
+	w.buf.PutByte(byte(indexVersion))
+	w.buf.PutBE64(indexPostingOffsetTable)
+	return w.f.Write(w.buf.Get())
+}
+
+func (w *binaryWriter) SymbolsWriter() io.Writer {
+	w.toc.Symbols = w.f.Pos()
+	return w
+}
+
+func (w *binaryWriter) PostingOffsetsWriter() io.Writer {
+	w.toc.PostingsOffsetTable = w.f.Pos()
+	return w
+}
+
+func (w *binaryWriter) WriteTOC() error {
+	w.buf.Reset()
+
+	w.buf.PutBE64(w.toc.Symbols)
+	w.buf.PutBE64(w.toc.PostingsOffsetTable)
+
+	w.buf.PutHash(w.crc32)
+
+	return w.f.Write(w.buf.Get())
+}
+
+func (w *binaryWriter) Write(p []byte) (int, error) {
+	n := w.f.Pos()
+	err := w.f.Write(p)
+	return int(w.f.Pos() - n), err
+}
+
+func (w *binaryWriter) Close() error {
+	return w.f.Close()
+}
+
+type postingValueOffsets struct {
+	offsets       []postingOffset
+	lastValOffset int64
+}
+
+type postingOffset struct {
+	// label value.
+	value string
+	// offset of this entry in posting offset table in index-header file.
+	tableOff int
+}
+
+type BinaryReader struct {
+	b   index.ByteSlice
+	toc *BinaryTOC
+
+	// Close that releases the underlying resources of the byte slice.
+	c io.Closer
+
+	// Map of LabelName to a list of some LabelValues's position in the offset table.
+	// The first and last values for each name are always present.
+	postings map[string]*postingValueOffsets
+	// For the v1 format, labelname -> labelvalue -> offset.
+	postingsV1 map[string]map[string]index.Range
+
+	symbols     *index.Symbols
+	nameSymbols map[uint32]string // Cache of the label name symbol lookups,
+	// as there are not many and they are half of all lookups.
+
+	dec *index.Decoder
+
+	version             int
+	indexVersion        int
+	indexLastPostingEnd int64
+}
+
+// NewBinaryReader loads or builds new index-header if not present on disk.
+func NewBinaryReader(ctx context.Context, logger log.Logger, bkt objstore.BucketReader, dir string, id ulid.ULID) (*BinaryReader, error) {
+	binfn := filepath.Join(dir, id.String(), block.IndexHeaderFilename)
+	br, err := newFileBinaryReader(binfn)
+	if err == nil {
+		return br, nil
+	}
+
+	level.Debug(logger).Log("msg", "failed to read index-header from disk; recreating", "path", binfn, "err", err)
+
+	start := time.Now()
+	if err := WriteBinary(ctx, bkt, id, binfn); err != nil {
+		return nil, errors.Wrap(err, "write index header")
+	}
+
+	level.Debug(logger).Log("msg", "built index-header file", "path", binfn, "elapsed", time.Since(start))
```

Comment:
Given we don't have any metric whenever an `index-header` is created, if there's a bug which re-creates the `index-header` for a block each time (because the reader detects it as corrupted) it would be very difficult to notice, unless you run Thanos with debug logging. Switching this log to `Info` may help.

Reply:
Right. In this case we need a metric (:

Reply:
Will add that in next PR. (: